docs/tutorials/sampling/ResamplingTutorial.py

```diff
@@ -0,0 +1,399 @@
+
+
+# Pick a starting point
+s = np.random.uniform(0, 1/n_particles)
+
+# Calculate N equidistant points from the starting point
+u_j = s + (1 / n_particles) * np.arange(n_particles)
+
+plot(normalised_weights, u_j)
+
+# %%
+# Stratified Resampler
+# ^^^^^^^^^^^^^^^^^^^^
+#
+# The :class:`~.StratifiedResampler` splits the whole CDF into :math:`N` evenly sized strata
+# (subpopulations). A random point is then chosen, independently, from each stratum. This results
+# in the gap between any two consecutive points being in the range :math:`(0, 2/N]`. The
+# Stratified resampler has a computational complexity of :math:`O(N)` where :math:`N` is the
+# number of resampled particles.
+#
+# An example of how the Stratified resampler picks points from the CDF is shown below. Black
+# lines represent the chosen points, with white dashed lines representing the strata boundaries.
+
+
+# Calculate lower bound of each stratum
+s_lb = np.arange(n_particles) * (1 / n_particles)
+
+# Independently pick a point from each stratum
+u_j = np.random.uniform(s_lb, s_lb + (1 / n_particles))
+
+plot(normalised_weights, u_j, s_lb)
+
+# %%
+# Preprocessing methods
+# ---------------------
+
+# %%
+# Residual Resampler
+# ^^^^^^^^^^^^^^^^^^
+#
+# The :class:`~.ResidualResampler` consists of two stages.
+#
+# **Stage 1**
+#
+# The first stage determines which particles have weight :math:`w^{i} \geq 1/N`, where
+# :math:`i \in 1, ..., N` denotes each particle. Each of these particles is then resampled
+# :math:`N^{i}_{j} = floor(Nw^{i}_{j})` times, where :math:`j=1` denotes stage 1. Hence,
+# :math:`N_1 = \sum_{i=1}^{N}N^i_1` represents the number of particles that are sampled in stage 1.
+# As these weights have been represented in the resampled set of particles, we are only interested
+# in the residual weights, left after the floor weights (:math:`N^{i}_{1}`) have been subtracted.
+#
+#
+# Reconsider our example with 5 particles, with weights shown by the plot below.
+
+plot(normalised_weights)
+
+# %%
+# As we're interested in particles of weight :math:`w^{i} \geq 1/N`, in this example we're looking
+# for particles with :math:`w^{i} \geq 1/5`. Hence particle 2 (orange, weight = 0.5) and particle
+# 5 (purple, weight = 0.1). We resample these particles :math:`N^{i}_{j} = floor(Nw^{i}_{j})`
+# times:
+#
+# - particle 2 gets resampled :math:`N^{2}_{1} = floor(Nw^{2}_{1}) = floor(5 * 0.5) = 2` times.
+# - particle 5 gets resampled :math:`N^{5}_{1} = floor(Nw^{5}_{1}) = floor(5 * 0.2) = 1` times.
+# - particles 1, 3 and 4 do not get resampled as their weights are :math:`\leq 1/5`.
+#
+#
+# A total of 3 particles were sampled from stage 1 (2x particle 2, 1x particle 5), hence
+# :math:`N_1 = \sum_{i=1}^{N}N^i_1 = 3`
+#
+#
+# Original CDF with integer weights removed:
+
+plot(normalised_weights, residual=True)
+
+
+# %%
+# New CDF:
+
+floors = np.floor(normalised_weights * n_particles)
+residuals = (normalised_weights - floors/n_particles)
+plot(residuals)
```

Comment:
So I think the confusion here is that if a particle has weight bigger than 1/N (N = no. particles), then as many chunks of 1/N as possible are subtracted from the weight, and the left over residual weight is carried on. E.g assuming 5 particles, a particle of weight 0.3 would be resampled once as 1/N = 0.2 can be subtracted only once from 0.3, leaving a residual weight of 0.1.

The residual weights are carried forward rather than the weight of the whole particle being removed.